### ISSUE TITLE
How to do Gradient Checkpoint + LoRA

### ISSUE THREAD

**Issue by tcapelle** [huggingface/peft#1142]
### System Info

<img width="570" alt="image" src="https://github.com/huggingface/peft/assets/18441985/9b3ae040-d78a-477b-a9ec-6ab26b687a68">

### Who can help?

I need help with using LoRA + gradient checkpointing.
Using the reentrant option appears to be the solution, but it slows down training a lot, for LLama-7b it's more than 2x the training time of a full fine-tune on the same hardware (A100).
<img width="817" alt="image" src="https://github.com/huggingface/peft/assets/18441985/6c58b8b2-eb3c-472a-8643-dcec6193dfe6">

We should be able to just use vanilla gradient checkpoint.

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model

# model_id, vocab = 'meta-llama/Llama-2-7b-hf', 32000
model_id, vocab = "stas/tiny-random-llama-2", 3000

seq_len = 1024
bs=8
use_lora=True

model_config = dict(
    pretrained_model_name_or_path=model_id,
    device_map=0,
    trust_remote_code=True,
    low_cpu_mem_usage=True,
    torch_dtype=torch.bfloat16,
    use_cache=False,
)

model = AutoModelForCausalLM.from_pretrained(**model_config)

# Just freeze embeddings for small memory decrease
model.model.embed_tokens.weight.requires_grad_(False);

if use_lora:
    lora_config = LoraConfig(
        r=2,  # the rank of the LoRA matrices
        lora_alpha=16, # the weight
        lora_dropout=0.1, # dropout to add to the LoRA layers
        bias="none", # add bias to the nn.Linear layers?
        task_type="CAUSAL_LM",
        target_modules=["q_proj", "k_proj","v_proj","o_proj"], # the name of the layers to add LoRA
    )
    
    model = get_peft_model(model, lora_config)

example = {"input_ids": torch.randint(0, vocab, size=(bs,seq_len), device="cuda:0"), 
           "labels":torch.randint(0, vocab, size=(bs,seq_len), device="cuda:0")}

import torch, peft, accelerate, transformers
for lib in [torch, peft, accelerate, transformers]:
    print(f"{lib.__name__}: {lib.__version__}")

model.train()
def call_forward():
    with torch.amp.autocast("cuda", dtype=torch.bfloat16):
        out = model(**example)
        loss = out.loss
    return loss

%timeit loss=call_forward()
loss=call_forward()
loss.requires_grad
# 5.48 ms ± 31.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# True

model.gradient_checkpointing_enable()
%timeit loss=call_forward()
loss=call_forward()
loss.requires_grad
# 5.13 ms ± 33.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# False

model.gradient_checkpointing_enable(dict(use_reentrant=False))
%timeit loss=call_forward()
loss=call_forward()
loss.requires_grad
# 7.23 ms ± 40.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# True
```

### Expected behavior

Nothing to add here.

**Comment by younesbelkada** [comment]
Hi @tcapelle 
Assuming you have a recent version of `transformers`, can you try to call `model.enable_input_require_grads()` after loading it ? (i.e. after `from_pretrained`) 
We do that automatically under the hood for quantized base models: https://github.com/huggingface/peft/blob/main/src/peft/utils/other.py#L105C17-L105C51 - but for now we require users to do it manually through that utility method

**Comment by BenjaminBossan** [comment]
Younes, I wonder if this shouldn't already be called automatically?

https://github.com/huggingface/peft/blob/9cdaed27693ec202b6792349cf41f22b394c8f5e/src/peft/peft_model.py#L396-L397

Or are we missing something?

**Comment by younesbelkada** [comment]
True, the issue here is that @tcapelle is calling `gradient_checkpointing_enable` after creating the `PeftModel` therefore it does not enter that condition: https://github.com/huggingface/peft/blob/9cdaed27693ec202b6792349cf41f22b394c8f5e/src/peft/peft_model.py#L125
So another fix could be to simply call `gradient_checkpointing_enable` before declaring the `PeftModel`

**Comment by tcapelle** [comment]
This actuall fixes the issue:
> calling model.enable_input_require_grads()` before calling `get_peft_model`
But training is still extremely slow

**Comment by tcapelle** [comment]
Do you know why we need to make the inputs require grad?

**Comment by tcapelle** [comment]
Actually the better fix is just using gradient checkpointing with `reentrant=False`

**Comment by tcapelle** [comment]
I am not convinced that adding require_grad to the input is a good idea. Do we have some analysis over this on memory consumption?
- reentrant = False seems a better option, but I am not sure we get the same results

**Comment by epinnock** [comment]
I'm currently trying to do this with accelerate, and I was wondering where should we be setting reentrant=False

**Comment by younesbelkada** [comment]
Hi @epinnock 
for setting `use_reentrant=False` you need to pass `gradient_checkpointing_kwargs={"use_reentrant": False}` when calling `model.gradient_checkpointing_enable()`

**Comment by AngryBacteria** [comment]
Hei there
So I think this issue is not yet solved, or at I am doing something wrong while trying to make Lora work. If I do not provide set the `gradient_checkpointing_kwargs={"use_reentrant": False}` value in the training arguments, the script just crashes. If I set the value the script works and runs, but the training is slowed down significantly compared to disabling gradient checkpointing. This is similar to what @tcapelle was reporting. The slowdown is in the realm of 2-3x and not the expected 1.2-1.3x. The slowdown is especially strong when using a higher batch size and/or sequence length. I do not see this behavior while doing full-finetuning, only with Lora. I would appreciate any help as I am not sure where I did something wrong.

Here is the link to my training script. I use the lora config. I tested batch size of 2 with 1024 sequence length. 
https://github.com/AngryBacteria/ba-gujen1-steij14/blob/main/research/training/train.py

**Comment by tobiapoppi** [comment]
Did someone ever notice that this solution is not working for loras on ViT layers but only on LLM layers?
I am using Qwen-2.5-VL-7B, everything frozen, lora on Vision layers, and lora on LLM layers.
without reentrant = True, loras on LLM gets gradients, and loras on Vision layers have grad None.
with reentrant = False problem get solved for lora on vision layers, but that's not a solution if grad is sharded across gpus with deepspeed. Still wondering why there is this discrepancy between loras on visual and textual architectures. if I get grad on LLM I should get it also on ViT. I don't have torch.no_grad or detach() or anything like that.

**Comment by BenjaminBossan** [comment]
@tobiapoppi Ideally, you could create a new issue and report your error there. Please provide as much info as possible (configs, hardware, relevant package versions) and a minimal reproducer, otherwise we're gonna have trouble reproducing the error.
